File: src/chap08_LinkedList/SwapNodesInPairs.java
package chap08_LinkedList;

public class SwapNodesInPairs {

    public static void main(String[] args) {
        ListNode<Integer> listNode6 = new ListNode<>(6);
        ListNode<Integer> listNode5 = new ListNode<>(5, listNode6);
        ListNode<Integer> listNode4 = new ListNode<>(4, listNode5);
        ListNode<Integer> listNode3 = new ListNode<>(3, listNode4);
        ListNode<Integer> listNode2 = new ListNode<>(2, listNode3);
        ListNode<Integer> listNode1 = new ListNode<>(1, listNode2);

        ListNode<Integer> swapped = swapPairsV2(listNode1);

        while(swapped != null) {
            System.out.println(swapped.val);
            swapped = swapped.next;
        }
    }

    private static <E extends Number> ListNode<E> swapPairsV1(ListNode<E> head) {
        ListNode<E> prev = new ListNode<>();
        ListNode<E> root = prev;
        prev.next = head;
        while (head != null) {
            ListNode<E> b = head.next;
            head.next = b.next;
            b.next = head;

            prev.next = b;

            head = head.next;
            prev = prev.next.next;
        }
        return root.next;
    }

    private static <E extends Number> ListNode<E> swapPairsV2(ListNode<E> head) {
        if (head != null) {
            ListNode<E> p = head.next;

            head.next = swapPairsV2(p.next);
            p.next = head;
            return p;
        }
        return head;
    }
}

File: src/chap08_LinkedList/SwapNodesInPairs.js
let swapPairs = function(head) {
    if (head !== null && head.next !== null) {
        let p = head.next;

        head.next = swapPairs(p.next);
        p.next = head;
        return p;
    }
    return head;
}

class Node {
    constructor(node, value) {
        this.next = node;
        this.val = value;
    }
}

let n6 = new Node(null, 6);
let n5 = new Node(n6, 5);
let n4 = new Node(n5, 4);
let n3 = new Node(n4, 3);
let n2 = new Node(n3, 2);
let n1 = new Node(n2, 1);

let swap = swapPairs(n1);
